homeassistant/helpers/entity.py

```diff
@@ -137,28 +137,28 @@ def name(self) -> Optional[str]:
         return None
 
     @property
-    def state(self) -> str:
+    def state(self) -> Optional[str]:
         """Return the state of the entity."""
```

Comment:
Is this really `Optional`, i.e. is it okay to return `None` here considering `STATE_UNKNOWN` exists and is the default? If yes, what's the difference between them?

Reply:
`None` will be converted to `STATE_UNKNOWN` when updating state, by the `Entity` base class.

Reply:
Sounds like a cleanup opportunity to me, would be better to have just one representation for unknown state everywhere. That's not a problem with this PR though.

Reply:
@scop Funny how you trigger on that part... the fact it has to be a string was my personal shocker... do you know how often we throw integers and floats at it? Sure, it is handled on writing the state later, but actually... yeah...

Reply:
@scop I agree actually... Maybe revert it and work towards changing the codebase?
The real shocker for me personally, it the fact it has to be a string... We throw everything at it from all our integrations (still handled, nevertheless, not intended).

Reply:
It's nice to not have to import `UNKNOWN_STATE` in every platform to represent unknown state. If we can make it more consistent in another way, that would be interesting.

Reply:
Personally, I would prefer an import of `UNKNOWN_STATE` over making things `Optional`, any time. In the long run, you would have to be sure to deal with those `NoneTypes` all the time.

Reply:
We only deal with this in one place, that's the point, I think.

Reply:
I get that, the result, however, is that we have to deal with checking it for a `None` value every time it is used elsewhere.

FYI, our docs also list `state` it is required:
https://developers.home-assistant.io/docs/en/entity_sensor.html

Reply:
The internal state of the entity shouldn't be used outside the entity. We use the state machine to access the state of the entities, in other places in Home Assistant. The `None` value is converted before the state machine is updated.

Reply:
Sure, but don't we restore states from it as well? Shouldn't it be all the same in that case?

Reply:
(PS: Kinda out of scope maybe, this PR was just to reflect/match the current state of affairs)

Reply:
The states to be saved are taken from the state machine, so we restore the state from the state machine. So you're right that that could mean a complication somewhere, if it's not handled by the restore state logic.

Reply:
I am all in favor of making it explicit and having it be `str` and setting `STATE_UNKNOWN` in the constructor if passed in state is None.

Reply:
ooh I realize now that this is helper/entity.py, not core.py. 

We convert it to a string and even with correct typing, i want to keep it that way, because I do not trust integrations, especially custom components. This will not be caught with just type checks.

Reply:
So since we cast it to a string, we could eh, do `Any`?

Reply:
I've applied a suggestion I got from @pvizeli:

`Union[None, str, int, float]`

Which, should cover our grounds for the entity helper. Based on the comments here, I've also adjusted the type hints for core entity. The discussion clearly settled on the state in the core is a string, so I've adjusted the core entity to reflect that.